File: src/protocolsession/ProtocolSessionList.cpp
//MIT License

//Copyright (c) 2020 bexoft GmbH (mail@bexoft.de)

//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:

//The above copyright notice and this permission notice shall be included in all
//copies or substantial portions of the Software.

//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//SOFTWARE.


#include "finalmq/protocolsession/ProtocolSessionList.h"
#include "finalmq/protocolsession/ProtocolSession.h"


namespace finalmq {


std::atomic_int64_t ProtocolSessionList::m_nextSessionId{1};


ProtocolSessionList::ProtocolSessionList()
{
}

ProtocolSessionList::~ProtocolSessionList()
{
}

std::int64_t ProtocolSessionList::addProtocolSession(IProtocolSessionPrivatePtr ProtocolSession, bool verified)
{
    std::unique_lock<std::mutex> lock(m_mutex);
    std::int64_t sessionId = m_nextSessionId.fetch_add(1);
    m_connectionId2ProtocolSession[sessionId] = { ProtocolSession, verified, {} };
    lock.unlock();
    return sessionId;
}

void ProtocolSessionList::removeProtocolSession(std::int64_t sessionId)
{
    std::unique_lock<std::mutex> lock(m_mutex);
    m_connectionId2ProtocolSession.erase(sessionId);
    lock.unlock();
}

std::vector< IProtocolSessionPrivatePtr > ProtocolSessionList::getAllSessions() const
{
    std::vector< IProtocolSessionPrivatePtr > protocolSessions;
    std::unique_lock<std::mutex> lock(m_mutex);
    protocolSessions.reserve(m_connectionId2ProtocolSession.size());
    for (auto it = m_connectionId2ProtocolSession.begin(); it != m_connectionId2ProtocolSession.end(); ++it)
    {
        if (it->second.verified)
        {
            protocolSessions.emplace_back(it->second.session);
        }
    }
    lock.unlock();
    return protocolSessions;
}

IProtocolSessionPtr ProtocolSessionList::getSession(std::int64_t sessionId) const
{
    std::unique_lock<std::mutex> lock(m_mutex);
    auto it = m_connectionId2ProtocolSession.find(sessionId);
    if (it != m_connectionId2ProtocolSession.end())
    {
        if (it->second.verified)
        {
            return it->second.session;
        }
    }
    return nullptr;
}


IProtocolSessionPrivatePtr ProtocolSessionList::findSessionByName(const std::string& sessionName) const
{
    IProtocolSessionPrivatePtr protocolSessions;
    std::unique_lock<std::mutex> lock(m_mutex);
    for (auto it = m_connectionId2ProtocolSession.begin(); it != m_connectionId2ProtocolSession.end(); ++it)
    {
        if (it->second.verified && it->second.name == sessionName)
        {
            return it->second.session;
        }
    }
    lock.unlock();
    return nullptr;
}


bool ProtocolSessionList::setSessionName(std::int64_t sessionId, const std::string& sessionName)
{
    bool ok = false;
    std::unique_lock<std::mutex> lock(m_mutex);
    auto it = m_connectionId2ProtocolSession.find(sessionId);
    if (it != m_connectionId2ProtocolSession.end())
    {
        if (!it->second.verified)
        {
            ok = true;
            it->second.name = sessionName;
            it->second.verified = true;
        }
    }
    return ok;
}



}   // namespace finalmq

File: csharp/src/protocolsession/ProtocolSessionList.cs
//MIT License

//Copyright (c) 2020 bexoft GmbH (mail@bexoft.de)

//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:

//The above copyright notice and this permission notice shall be included in all
//copies or substantial portions of the Software.

//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//SOFTWARE.


namespace finalmq
{

    internal interface IProtocolSessionList
    {
        long AddProtocolSession(IProtocolSessionPrivate protocolSession, bool verified);
        void RemoveProtocolSession(long sessionId);
        IList<IProtocolSessionPrivate> GetAllSessions();
        IProtocolSession GetSession(long sessionId);
        IProtocolSessionPrivate? FindSessionByName(string sessionName);
        bool SetSessionName(long sessionId, string sessionName);
    }

    internal class ProtocolSessionList : IProtocolSessionList
    {
        public ProtocolSessionList()
        {
        }

        public long AddProtocolSession(IProtocolSessionPrivate protocolSession, bool verified)
        {
            long sessionId = 0;
            lock (m_mutex)
            {
                sessionId = Interlocked.Increment(ref m_nextSessionId);
                m_connectionId2ProtocolSession[sessionId] = new SessionData(protocolSession, verified, "");
            }
            return sessionId;
        }
        public void RemoveProtocolSession(long sessionId)
        {
            lock (m_mutex)
            {
                m_connectionId2ProtocolSession.Remove(sessionId);
            }
        }
        public IList<IProtocolSessionPrivate> GetAllSessions()
        {
            IList<IProtocolSessionPrivate> protocolSessions = new List<IProtocolSessionPrivate>();
            lock (m_mutex)
            {
                foreach (var entry in m_connectionId2ProtocolSession)
                {
                    SessionData sessionData = entry.Value;
                    if (sessionData.verified)
                    {
                        protocolSessions.Add(sessionData.session);
                    }
                }
            }
            return protocolSessions;
        }
        public IProtocolSession GetSession(long sessionId)
        {
            lock (m_mutex)
            {
                SessionData? sessionData = null;
                m_connectionId2ProtocolSession.TryGetValue(sessionId, out sessionData);
                if (sessionData != null)
                {
                    return sessionData.session;
                }
            }
            throw new System.Collections.Generic.KeyNotFoundException("Session ID " + sessionId + " not found");
        }
        public IProtocolSessionPrivate? FindSessionByName(string sessionName)
        {
            lock (m_mutex)
            {
                foreach (var entry in m_connectionId2ProtocolSession)
                {
                    if (entry.Value.verified && entry.Value.name == sessionName)
                    {
                        return entry.Value.session;
                    }
                }
            }
            return null;
        }

        public bool SetSessionName(long sessionId, string sessionName)
        {
            bool ok = false;
            lock (m_mutex)
            {
                SessionData? sessionData = null;
                m_connectionId2ProtocolSession.TryGetValue(sessionId, out sessionData);
                if (sessionData != null && !sessionData.verified)
                {
                    ok = true;
                    sessionData.name = sessionName;
                    sessionData.verified = true;
                }
            }
            return ok;
        }

        class SessionData
        {
            public SessionData(IProtocolSessionPrivate session, bool verified, string name)
            {
                this.session = session;
                this.verified = verified;
                this.name = name;
            }
            public IProtocolSessionPrivate session;
            public bool verified = false;
            public string name;
        };

        private IDictionary<long, SessionData>  m_connectionId2ProtocolSession = new Dictionary<long, SessionData>();
        private static long                     m_nextSessionId = 1;
        private readonly object                 m_mutex = new object();
    }

}   // namespace finalmq
